concordium-consensus/src/Concordium/KonsensusV1/Consensus.hs

```diff
@@ -0,0 +1,210 @@
+
+-- |Consensus V1
+module Concordium.KonsensusV1.Consensus where
+
+import Control.Monad.State
+import Data.Maybe (fromMaybe, isJust)
+
+import qualified Data.Vector as Vector
+import Lens.Micro.Platform
+
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.Parameters
+import Concordium.Types.Transactions
+import Concordium.Types.Updates (uiHeader, uiPayload, updateType)
+import Concordium.Utils
+
+import Concordium.GlobalState.BlockState
+import qualified Concordium.GlobalState.Persistent.BlockState as PBS
+import qualified Concordium.GlobalState.TransactionTable as TT
+import qualified Concordium.GlobalState.Types as GSTypes
+import Concordium.KonsensusV1.TransactionVerifier
+import Concordium.KonsensusV1.TreeState.Implementation
+import Concordium.KonsensusV1.TreeState.Types
+import Concordium.KonsensusV1.Types
+import Concordium.Scheduler.Types (updateSeqNumber)
+import Concordium.TimeMonad
+import qualified Concordium.TransactionVerification as TVer
+
+-- |Result of attempting to add a 'BlockItem' into tree state.
+data AddBlockItemResult
+    = -- |The transaction was accepted into
+      -- the tree state.
+      Accepted
+    | -- |The transaction was rejected. See
+      -- the 'TVer.VerificationResult' for why it
+      -- was rejected.
+      Rejected !TVer.VerificationResult
+    | -- |The transaction was already present
+      -- in the tree state.
+      Duplicate
+    | -- |The transaction nonce was obsolete,
+      -- i.e. inferior to the next available nonce.
+      Obsolete
+    deriving (Eq, Show)
+
+-- |Adds a transaction into the pending transaction table
+-- if it's eligible.
+-- A transaction is eligible if the nonce of the transaction is at least
+-- as high as the next available nonce observed with respect to the focus block.
+--
+-- We always check that the nonce of the transaction (if it has origin 'Block') is
+-- at least what is recorded with respect to next available nonce for the
+-- account. This is not necessary for transactions received individually
+-- as they have already been pre-verified, and they will get rejected if they
+-- do not yield exactly the recorded next available nonce.
+--
+-- This is an internal function only and should not be called directly.
+addPendingTransaction ::
+    ( MonadState (SkovData (MPV m)) m,
+      TimeMonad m,
+      BlockStateQuery m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState (MPV m)
+    ) =>
+    -- |Origin of the transaction.
+    TransactionOrigin ->
+    -- |The transaction.
+    BlockItem ->
+    m ()
+addPendingTransaction origin bi = do
+    case wmdData bi of
+        NormalTransaction tx -> do
+            fbState <- bpState <$> (_focusBlock <$> gets' _skovPendingTransactions)
+            macct <- getAccount fbState $! transactionSender tx
+            nextNonce <- fromMaybe minNonce <$> mapM (getAccountNonce . snd) macct
+            when (nextNonce <= transactionNonce tx || origin == Individual) $ do
```

Comment:
yes that's a mistake.

Reply:
Edit: I just looked back at this. I believe it was correct as it was stated before. 
Well, the code was the documentation was wildly misleading. 

I have reverted my change and improved the documentation. Please verify.

Reply:
Due to the precondition of this function i.e. 
`Pre condition: A transaction must've been pre-verified prior to being called here.`
Then it will not violate the the pending transaction table & focus block as the verification (when the transaction is received individually) ensures that the transaction nonce is exactly the next available nonce for the account. 

So when the transaction is received individually, then this check is not needed at this place.
This should  always be the case as the documentation of the function clearly states: 
`-- This is an internal function only and should not be called directly.`